src/data_access/FileAPIDataAccessObject.java

```diff
@@ -0,0 +1,90 @@
+
+import java.io.*;
+import java.time.LocalDateTime;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class FileAPIDataAccessObject implements SpotifyAPIDataAccessInterface {
+    private final File csvFile;
+    private final Map<String, Integer> headers = new LinkedHashMap<>();
+    private String authToken;
+    private LocalDateTime authLastCreated;
+    private String clientId;
+    private String clientSecret;
+
+    public FileAPIDataAccessObject(String csvPath) throws IOException {
+        csvFile = new File(csvPath);
+        headers.put("creation_time", 0);
+        headers.put("auth_token", 1);
+        headers.put("client_id", 2);
+        headers.put("client_secret", 3);
+
+        if(this.csvFile.length() == 0){
+            save(); // create file but still need to manually add client id and secret for program to run
+        }else{
+            try (BufferedReader reader = new BufferedReader(new FileReader(csvFile))) {
+                String header = reader.readLine();
+
+                assert header.equals("creation_time,auth_token,client_id,client_secret");
+
+                String row =  reader.readLine();
+
+                assert row != null; // should always contain client_id and cleint_secret
+
+                String[] col = row.split(",");
+                clientId = String.valueOf(col[headers.get("client_id")]);
+                clientSecret = String.valueOf(col[headers.get("client_secret")]);
+                String authText = String.valueOf(col[headers.get("auth_token")]);
+                if(!Objects.equals(authText, "null")) {
+                    authToken = authText;
+                }
+                String creationTimeText = String.valueOf(col[headers.get("creation_time")]);
+                if(!Objects.equals(creationTimeText, "null")){
+                    authLastCreated = LocalDateTime.parse(creationTimeText);
+                }
+            }
+        }
+    }
+    private void save(){
+        BufferedWriter writer;
+        try {
+            writer = new BufferedWriter(new FileWriter(csvFile));
+            writer.write(String.join(",", headers.keySet()));
+            writer.newLine();
+            String line = String.format("%s,%s,%s,%s",authLastCreated,authToken,clientId,clientSecret);
+            writer.write(line);
+            writer.newLine();
+            writer.close();
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    public String getAuthToken(){
+        return this.authToken;
+    }
+    public String getClientId(){
+        return this.clientId;
+    }
+    public String getClientSecret(){
+        return this.clientSecret;
+    }
+    @Override
+    public boolean expiredToken() {
+        LocalDateTime now = LocalDateTime.now();
+        if(authToken == null) return true;
+        return authLastCreated.plusSeconds(3600).isBefore(now);
+    }
```

Comment:
I think this should probably be fine for our purposes but it is technically possible that this function returns `true` and then we make a request to the Spotify API and our access token expires before the request is processed on their side (e.g. we confirm that a token expiring in 1 second is okay to use so we send a request to Spotify to get a track, but our internet connection isn't great so it takes a couple of seconds, which leads to our request being denied because our access token is now expired, and then our program crashes).

At a later date, we could consider adding code that checks the response of any Spotify request to see if we get a specific error stating that our token is invalid. If this is the case, then we fetch a new access token and queue the same request again.

Reply:
Great point, we can call the api to get songs and if the response says its invalid we call the gnereate auth token func. Will add that now